### ISSUE TITLE
Generated text is empty when a tool is invoked.

### ISSUE THREAD

**Issue by mbokinala** [vercel/ai#1831]
### Description

When using the code sample on the [tools documentation page](https://sdk.vercel.ai/docs/ai-sdk-core/tools-and-tool-calling#tools-and-tool-calling), (see code sample below) the generated text output is an empty string. This is the output:

```json
{
  "text": "",
  "toolCalls": [
    {
      "type": "tool-call",
      "toolCallId": "call_V0T4bn4INUM9Rcqjpzs0DKiJ",
      "toolName": "weather",
      "args": {
        "location": "San Francisco"
      }
    }
  ],
  "toolResults": [
    {
      "toolCallId": "call_V0T4bn4INUM9Rcqjpzs0DKiJ",
      "toolName": "weather",
      "args": {
        "location": "San Francisco"
      },
      "result": {
        "location": "San Francisco",
        "temperature": 79
      }
    }
  ],
  "finishReason": "tool-calls",
  "usage": {
    "promptTokens": 68,
    "completionTokens": 14,
    "totalTokens": 82
  },
  "warnings": [],
  "rawResponse": {
    "headers": {
      "alt-svc": "h3=\":443\"; ma=86400",
      "cf-cache-status": "DYNAMIC",
      "cf-ray": "88eb0725f84315e3-SJC",
      "connection": "keep-alive",
      "content-encoding": "br",
      "content-type": "application/json",
      "date": "Tue, 04 Jun 2024 21:31:08 GMT",
      "openai-organization": "user-lcxfrg25h6hcz9xkakncaeaw",
      "openai-processing-ms": "1663",
      "openai-version": "2020-10-01",
      "server": "cloudflare",
      "set-cookie": "_cfuvid=9xHXI8U1g30WIBWUYIMWPQ95_tNdzhNp3VmVV.HNoy0-1717536668421-0.0.1.1-604800000; path=/; domain=.api.openai.com; HttpOnly; Secure; SameSite=None",
      "strict-transport-security": "max-age=15724800; includeSubDomains",
      "transfer-encoding": "chunked",
      "x-ratelimit-limit-requests": "500",
      "x-ratelimit-limit-tokens": "30000",
      "x-ratelimit-remaining-requests": "499",
      "x-ratelimit-remaining-tokens": "29964",
      "x-ratelimit-reset-requests": "120ms",
      "x-ratelimit-reset-tokens": "72ms",
      "x-request-id": "req_bc5f983ee98ba53bd72ca43d78f41325"
    }
  },
  "responseMessages": [
    {
      "role": "assistant",
      "content": [
        {
          "type": "text",
          "text": ""
        },
        {
          "type": "tool-call",
          "toolCallId": "call_V0T4bn4INUM9Rcqjpzs0DKiJ",
          "toolName": "weather",
          "args": {
            "location": "San Francisco"
          }
        }
      ]
    },
    {
      "role": "tool",
      "content": [
        {
          "type": "tool-result",
          "toolCallId": "call_V0T4bn4INUM9Rcqjpzs0DKiJ",
          "toolName": "weather",
          "result": {
            "location": "San Francisco",
            "temperature": 79
          }
        }
      ]
    }
  ]
}
```

### Code example

```ts
import { z } from 'zod';
import { generateText, tool } from 'ai';
import { openai } from '@ai-sdk/openai';

async function main() {
  const result = await generateText({
    model: openai('gpt-4-turbo'),
    tools: {
      weather: tool({
        description: 'Get the weather in a location',
        parameters: z.object({
          location: z.string().describe('The location to get the weather for'),
        }),
        execute: async ({ location }) => ({
          location,
          temperature: 72 + Math.floor(Math.random() * 21) - 10,
        }),
      }),
    },
    prompt:
      'What is the weather in San Francisco and what attractions should I visit?',
  });


  console.log(JSON.stringify(result, null, 2))
}

main();
```

**Comment by nicoalbanese** [comment]
Hi @mbokinala! This is because the AI SDK doesn't automatically send the tool call result back to the model as that would be add an additional call. If you would like that behaviour, you can add `maxToolRoundtrips: 3` to the config object below prompt (play around with the number here, just make sure it's above 1), and then it will automatically send the tool call result back to the model.

**Comment by lgrammel** [comment]
I've added a `tool roundtrips` section to the tools page.

**Comment by lgrammel** [comment]
@JasonDevTech which `@ai-sdk/anthropic` version are you using?

**Comment by lgrammel** [comment]
@JasonDevTech that looks like a bug in our anthropic provider. taking a look

**Comment by JasonDevTech** [comment]
It doesn't seem like a bug since it's in both OpenAI and Anthropic providers. I'm using the latest version on NPM for both providers.

**Comment by JasonDevTech** [comment]
Here's a snippet of the issue below. As you can see, there's always a block of empty text.
{
    role: 'assistant',
    content: [
      { type: 'text', text: '' },
      {
        type: 'tool_use',
        id: 'toolu_01DEAC4NCkFvBJXzdNqcUrJf',
        name: 'image',
        input: {
          query: 'some image'
        }
      }
    ]
  },
  {
    role: 'user',
    content: [
      {
        type: 'tool_result',
        tool_use_id: 'toolu_01DEAC4NCkFvBJXzdNqcUrJf'

**Comment by lgrammel** [comment]
@JasonDevTech are you saying that you are seeing this error with OpenAI as well?

**Comment by lgrammel** [comment]
the issue is on provider level. the anthropic provider should filter empty text blocks

**Comment by JasonDevTech** [comment]
OpenAI allows empty blocks of content which is why there is no error (but there is still empty text blocks) - however Anthropic does not. In general, on the provider level (for all providers), it should filter out empty text blocks.

**Comment by lgrammel** [comment]
@JasonDevTech agreed, that's the better approach. thanks!

**Comment by lgrammel** [comment]
https://github.com/vercel/ai/pull/1874

**Comment by lgrammel** [comment]
@JasonDevTech [ai@3.1.30](https://github.com/vercel/ai/releases/tag/ai%403.1.30) should fix this

**Comment by JasonDevTech** [comment]
Thanks! Will try now

**Comment by flexchar** [comment]
Hey @lgrammel, how about using `maxToolRoundtrips` in `streamText` method?

**Comment by baptisteArno** [comment]
> Hey @lgrammel, how about using `maxToolRoundtrips` in `streamText` method?

#1943

**Comment by FakieHeelflip** [comment]
There is no maxToolRoundtrips parameter in the generateText() function?

**Comment by flexchar** [comment]
> There is no maxToolRoundtrips parameter in the generateText() function?

It's been removed since version `4.0`. Please see https://ai-sdk.dev/docs/migration-guides/migration-guide-4-0#remove-roundtrips. :)

**Comment by FakieHeelflip** [comment]
> > There is no maxToolRoundtrips parameter in the generateText() function?
> 
> It's been removed since version `4.0`. Please see https://ai-sdk.dev/docs/migration-guides/migration-guide-4-0#remove-roundtrips. :)

I am on version `5.0` and therefore there is no mention of it in the docs or migration guides.: https://ai-sdk.dev/docs/migration-guides/migration-guide-5-0 :(

**Comment by nqdai1992** [comment]
I am encountering an issue with the Tavily tool. Now, I need to make two calls to perform a search, which is not a good idea. Does anyone have a better solution? By the way, in version 5.0, we use `stopWhen: stepCountIs(3)` instead of `maxToolRoundtrips`.

**Comment by lgrammel** [comment]
## Identify issue type

Identified as **Bug**  
Confidence: **High**

### Rationale
Although the original empty final text was explained as expected without additional tool roundtrips, comments identified a concrete provider failure caused by empty text blocks, which maintainers fixed as a bug.

### Evidence
- Tool calls produced empty assistant text blocks alongside the tool-use content.
- During tool roundtrips, Anthropic rejected those blocks with “text content blocks must be non-empty.”
- A maintainer agreed empty blocks should be filtered, linked PR #1874, and released ai@3.1.30 as a fix.
